### ISSUE TITLE
Automated play improvements

### ISSUE THREAD

**Issue by seamuslowry** [seamuslowry/hundred-and-ten#153]
Build an AI training system for my hundred-and-ten card game that can train computer players using reinforcement learning. The AI should:

1. Train entirely offline through self-play (no network calls)
2. Be packaged as a separate Python package that consumers can use with hundred-and-ten
3. Eventually support easy/medium/hard difficulty levels (but optimal play for now is fine)
4. Export trained models that can be bundled and run locally

### Phase 1: Analysis & Setup
1. **Analyze the existing codebase** to understand:
   - Game state representation
   - Player interface and how moves are made
   - Game loop and turn structure
   - How bidding, trump selection, discarding, and card playing work
   - Current player implementations (if any)

2. **Create new package structure**:
   ```
   hundred-and-ten-ai/
   ├── pyproject.toml
   ├── src/hundred_and_ten_ai/
   │   ├── environment/
   │   ├── training/
   │   ├── agents/
   │   └── inference/
   ├── models/
   └── scripts/
   ```

### Phase 2: Environment Wrapper
3. **Build Gym-compatible environment** (`environment/gym_wrapper.py`):
   - Wrap the existing game engine
   - Define observation space (game state encoding)
   - Define action space (all possible moves)
   - Implement step(), reset(), render() methods
   - Handle partial observability (each player sees their own cards)

4. **Create state encoder** (`environment/state_encoder.py`):
   - Convert game state to neural network input format
   - Encode: hand cards, trump suit, current trick, scores, bid info, etc.
   - Should be a fixed-size vector representation

### Phase 3: Reward Structure
5. **Implement reward shaping** (`training/reward_shaping.py`):
   - **Episode termination**: End after N hands or when score threshold reached
   - **Reward components**:
     - Winning/losing the game (final outcome)
     - Meeting/failing bid contracts
     - Winning tricks
     - Relative performance vs opponents
   - Handle negative scoring issue with proper normalization

### Phase 4: Training Infrastructure
6. **Set up self-play training** (`training/self_play.py`):
   - Use Stable-Baselines3 (PPO or A2C algorithm)
   - Configure hyperparameters
   - Implement multi-agent training (4 players learning simultaneously)
   - Add training callbacks for monitoring progress

7. **Create training script** (`scripts/train.py`):
   - CLI for starting training runs
   - Save checkpoints periodically
   - Log metrics (win rate, average score, convergence)

### Phase 5: Agent Implementations
8. **Build agent interfaces** (`agents/`):
   - `base_agent.py`: Abstract interface for all agents
   - `trained_agent.py`: Loads and uses trained RL models
   - `random_agent.py`: Baseline random player for testing
   - Ensure agents implement the same interface as existing players

9. **Create inference wrapper** (`inference/ai_player.py`):
   - Production-ready AI player class
   - Loads trained model from file
   - Provides clean API: `bid()`, `select_trump()`, `discard()`, `play()`
   - No network calls, runs entirely locally

### Phase 6: Training & Validation
10. **Train initial model**:
    - Start with shorter training runs to validate setup
    - Monitor for convergence
    - Test against random baseline
    - Train longer for production model

11. **Export trained models**:
    - Save final model weights
    - Create model metadata (version, training config, performance metrics)
    - Document how to load and use models

### Phase 7: Integration & Documentation
12. **Make it usable by hundred-and-ten**:
    - Update hundred-and-ten to optionally depend on hundred-and-ten-ai
    - Provide example of using AI player in game
    - Ensure it works with existing game code

13. **Documentation**:
    - README with training instructions
    - API documentation for using trained agents
    - Notes on extending to easy/medium/hard variations

## Key Technical Decisions to Make

While implementing, please consider:
- **State representation**: What's the most efficient encoding of game state?
- **Action masking**: How to handle invalid moves (can't play card not in hand)?
- **Partial observability**: Each player only sees their cards - how to handle this in training?
- **Training opponents**: Should we train against other learning agents, or mix of learning + random?
- **Model architecture**: Simple feedforward network or something more complex?

## Dependencies to Add

- `gymnasium` (modern OpenAI Gym)
- `stable-baselines3` (RL algorithms)
- `torch` or `tensorflow` (neural networks)
- Your existing hundred-and-ten package

## Success Criteria

The system is complete when:
1. ✅ Training runs successfully through self-play
2. ✅ Trained agent beats random player consistently
3. ✅ Model can be saved and loaded
4. ✅ AI player integrates cleanly with existing game code
5. ✅ No network calls required during inference
6. ✅ Clear path to creating difficulty variations
